### PR TITLE
Introduce CF-IL implementation

### PR DESCRIPTION
**Pull request by agrudkow**

### CODE REVIEW

File: cf_il/model.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+        return loss.item()  # type: ignore[no-any-return]
+
+    def recover_memory(
+        self,
+        num_tasks: int,
+        num_classes_per_task: int = 2,
+        num_images_per_class: int = 10,
+        scale: float = 1,
+    ) -> None:
```
Comment (agrudkow) [inline cf_il/model.py]: Ustawić model w tryb ewaluacji, a następnie wrócić do poprzedniego staniu (tak jak jest to robione w funkcjach `evaluate` w `training.py`), bo tak to chyba będziemy aktualizować wagi podczas destylacji wiedzy.

File: cf_il/recover_memory.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        pseudo_labels = generate_dirichlet(
+            batch_size=num_images_per_class,
+            class_id=classes,
+            scale=scale,
+            similarity_matrix=norm_similarity_matrix,
+        )
+        for i in range(num_images_per_class):
+            rand_img = np.random.normal(0.5, 0.5, size=(32, 32, 3))
+            rand_img = rand_img.astype(np.float32)
+            rand_img_tensor = torch.tensor(rand_img.reshape((1, 3, 32, 32)), requires_grad=True)
```
Comment (agrudkow) [inline cf_il/recover_memory.py]: rozmiar obrazka, jest raczej zależny od zbioru danych, także trzeba to chyba sparametryzować

File: cf_il/recover_memory.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+            logits = model(rand_img_tensor)
+
+            loss = criterion(logits, torch.tensor(pseudo_labels[i]).reshape(1, -1))
+            loss.backward()
```
Comment (agrudkow) [inline cf_il/recover_memory.py]: czy nie powinniśmy wyżej zerować gradientu?

File: cf_il/model.py (hunk 1 above)
Comment (agrudkow) [reply cf_il/model.py]: EDIT: chyba jednak nie, bo nie robimy kroku optymatyzatora (ale wciąż przejście w tryb eval chyba może zwiększyć szybkość destylacji)

File: cf_il/train.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    results_mask_classes: List[List[float]] = []
+
+    model_stash = create_stash(model, args, dataset)
+
+    if args.tensorboard:
+        tb_logger = TensorboardLogger(args, dataset.SETTING, model_stash)
+        model_stash['tensorboard_name'] = tb_logger.get_name()
+
+    for task_idx in range(num_tasks):
+        model.net.train()
```
Comment (agrudkow) [inline cf_il/train.py:49]: to chyba można przenieść poza pętlę, nie ma sensu tego wywoływać za każdym razem

File: cf_il/recover_memory.py (hunk 2 above)
Comment (ptomaszewska) [reply cf_il/recover_memory.py]: To prawda :) Zwłaszcza, że drugi zbiór, którego używali w artykule: Tiny ImageNet ma rozmiar 64x64. Można chyba po prostu dodać info o rozmiarze jako kolejny parametr tej funkcji

File: scripts/test_recovery_memeory.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    model=model,
+    num_tasks=cif.N_TASKS,
+    num_images_per_class=10,
+)[0][0].reshape(32, 32, 3)
+
+print(img_noise.shape)
+#img_noise=img_noise.astype(np.float64)
+plt.imshow(transforms.ToPILImage()(img_noise.astype('uint8')))  # unit8 jest dziwny bez niego wyżej działa
+# https://stackoverflow.com/questions/62617533/trouble-with-pytorchs-topilimage
+plt.show()
```
Comment (ptomaszewska) [inline scripts/test_recovery_memeory.py]: Być może problem z wyświetlaniem obrazków wynika z tego, że wartości pikseli nie zawsze po dodaniu gradientów są z przedziału (0,1) i należałoby wcześniej zrobić clipping?

Comment (ptomaszewska) [reply scripts/test_recovery_memeory.py]: jest mała literówka w nazwie skryptu

File: cf_il/train.py (hunk 4 above)
Comment (ptomaszewska) [reply cf_il/train.py:49]: Nie do końca rozumiem co chcesz przenieść poza pętlę.

File: cf_il/recover_memory.py (hunk 3 above)
Comment (ptomaszewska) [reply cf_il/recover_memory.py]: Moim zdaniem na wszelki wypadek po loss.backward() warto dodać zerowanie, bo podobno domyślnie Pytorch kumuluje gradienty. Zazwyczaj gradienty zeruje się korzystając z optymalizatora, a my go tutaj w ogóle nie potrzebujemy... Znalazłam jednak taką komendę stosowaną w przypadku DeepDream, którą powinno się zastosować po dodaniu do szumu gradientu: rand_img_tensor.grad.data.zero_(), ale nie jestem pewna czy ona się u nas sprawdzi.

File: cf_il/model.py (hunk 1 above)
Comment (ptomaszewska) [reply cf_il/model.py]: Wydaje się, że chyba jest dobrze. Wagi będą aktualizowane tylko raz na podstawie połączonej straty z klasyfikacji na prawdziwych i destylowanych obrazkach.

File: scripts/test_recovery_memeory.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+setattr(args, "validation", True)
+setattr(args, "batch_size", 32)
+
+# Load dataset
+cif = SequentialCIFAR10(args)
+
+num_classes = cif.N_CLASSES_PER_TASK * cif.N_TASKS
+
+# Load model
+model = torch.hub.load('pytorch/vision:v0.10.0', 'resnet18', pretrained=True)  # False
```
Comment (ptomaszewska) [inline scripts/test_recovery_memeory.py]: w artykule nie używali pretrenowanych sieci

File: cf_il/model.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        self.buffer.empty()
+
+        synth_images, synth_logits = rec_mem(
+            model=self.net,
```
Comment (ptomaszewska) [inline cf_il/model.py:89]: czy tutaj nie powinny być generowane obrazki tylko z klas, które model już widział? Mam wrażenie, że teraz generuje z wszystkich.

File: cf_il/main.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    required=True,
+    help='The batch size of the memory buffer.',
+)
+
+# CF-IL
+parser.add_argument(
+    '--alpha',
+    type=float,
+    required=True,
+    help='Penalty weight.',
```
Comment (ptomaszewska) [inline cf_il/main.py]: może lepiej w opisie dać "degree of distilation" jak jest w artykule?

File: cf_il/main.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+)
+
+
+def fake_args(args: Namespace) -> Namespace:
```
Comment (ptomaszewska) [inline cf_il/main.py]: chyba ta funkcja nie jest potem używana - dlaczego?

File: cf_il/model.py (hunk 1 above)
Comment (agrudkow) [reply cf_il/model.py]: True, choć `eval` robi jeszcze kilka rzeczy z tego co wyczytałem https://stackoverflow.com/questions/60018578/what-does-model-eval-do-in-pytorch

File: cf_il/train.py (hunk 4 above)
Comment (agrudkow) [reply cf_il/train.py:49]: chodziło mi aby linię 49 przenieść do 47, bo tak za każdym razem to wołamy (choć do póki nie ma nigdzie ustawiania na val to można to zrobić, ale jak już będzie to musi tak zostać, także w sumie lepiej jak zostanie xd)

File: scripts/test_recovery_memeory.py (hunk 5 above)
Comment (agrudkow) [reply scripts/test_recovery_memeory.py]: > Być może problem z wyświetlaniem obrazków wynika z tego, że wartości pikseli nie zawsze po dodaniu gradientów są z przedziału (0,1) i należałoby wcześniej zrobić clipping?

możliwe

> jest mała literówka w nazwie skryptu

racja, wkradł się babol

File: scripts/test_recovery_memeory.py (hunk 6 above)
Comment (agrudkow) [reply scripts/test_recovery_memeory.py]: True, zostawiłem to testowo, ale trzeba zmienić

File: cf_il/model.py (hunk 7 above)
Comment (agrudkow) [reply cf_il/model.py:89]: racja, jest tu duży błąd

File: cf_il/main.py (hunk 8 above)
Comment (agrudkow) [reply cf_il/main.py]: Faktycznie, bardziej to pasuje

File: cf_il/main.py (hunk 9 above)
Comment (agrudkow) [reply cf_il/main.py]: na początku chciałem jej użyć aby sobie mock-ować argumety, ale koniec końców nie użyłem jej, więc jest do wywalenia

File: cf_il/recover_memory.py (hunk 3 above)
Comment (agrudkow) [reply cf_il/recover_memory.py]: Jednak nie ma potrzeby resetowania gradientu, bo za każdym razem tworzymy nowy tensor (a przynajmniej tak mi się wydaje)
